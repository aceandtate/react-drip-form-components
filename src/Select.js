import React from 'react';
import { dripFormField, FieldPropTypes } from 'react-drip-form';
import ErrorLabel from './internal/ErrorLabel';
import Spinner from './internal/Spinner';
import SelectField from './internal/SelectField';
import SelectControl from './internal/SelectControl';
import defaultProps from './internal/defaultFieldProps';


export const Select = (props) => {
  const {
    input,
    props: {
      shouldDisplayError,
      shouldDisplaySpinner,
      children,
      ...rest
    },
    meta,
  } = props;

  const displayError = shouldDisplayError(props);
  const displaySpinner = shouldDisplaySpinner(props);

  return (
    <SelectField showSpinner={displaySpinner}>
      <SelectControl
        {...rest}
        {...input}
        error={displayError}
      >
        {children}
      </SelectControl>

      {displayError &&
        <ErrorLabel>{meta.error}</ErrorLabel>
      }

      {displaySpinner &&
        <Spinner />
      }
    </SelectField>
  );
};

Select.propTypes = FieldPropTypes;


export default dripFormField('select', {
  defaultProps,
})(Select);
