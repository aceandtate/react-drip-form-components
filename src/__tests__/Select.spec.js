import React from 'react';
import { shallow } from 'enzyme';
import { mockFieldProps } from './utils';
import { Select } from '../Select';


describe('<Select />', () => {
  test('Should be pass function props', () => {
    const shouldDisplayError = jest.fn();
    const shouldDisplaySpinner = jest.fn();
    const props = mockFieldProps({
      shouldDisplayError,
      shouldDisplaySpinner,
    });

    expect(shouldDisplayError.mock.calls.length).toBe(0);
    expect(shouldDisplaySpinner.mock.calls.length).toBe(0);

    shallow(<Select {...props} />);

    expect(shouldDisplayError.mock.calls.length).toBe(1);
    expect(shouldDisplayError.mock.calls[0]).toEqual([props]);
    expect(shouldDisplaySpinner.mock.calls.length).toBe(1);
    expect(shouldDisplaySpinner.mock.calls[0]).toEqual([props]);
  });


  test('Should be render control', () => {
    const props = mockFieldProps();

    const wrapper = shallow(
      <Select {...props}>
        select field
      </Select>
    );

    expect(wrapper).toMatchSnapshot();
  });


  test('Should be render disabled control', () => {
    const props = mockFieldProps({
      disabled: true,
    });

    const wrapper = shallow(<Select {...props} />);

    expect(wrapper).toMatchSnapshot();
  });


  test('Should be render error text', () => {
    const props = mockFieldProps({
      shouldDisplayError: () => true,
      meta: {
        error: 'error!!',
      },
    });

    const wrapper = shallow(<Select {...props} />);

    expect(wrapper).toMatchSnapshot();
  });


  test('Should be render spinner', () => {
    const props = mockFieldProps({
      shouldDisplaySpinner: () => true,
    });

    const wrapper = shallow(<Select {...props} />);

    expect(wrapper).toMatchSnapshot();
  });
});
