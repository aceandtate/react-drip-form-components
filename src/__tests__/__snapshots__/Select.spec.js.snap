// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Select /> Should be render control 1`] = `
<styled.div
  showSpinner={false}
  theme={
    Object {
      "rdf": Object {
        "background": "rgb(255, 255, 255)",
        "backgroundDisabled": "rgb(245, 245, 245)",
        "backgroundEnter": "rgb(255, 255, 255)",
        "backgroundError": "rgb(255, 255, 255)",
        "backgroundFocus": "rgb(255, 255, 255)",
        "borderColor": "rgb(219, 219, 219)",
        "borderColorDisabled": "rgb(245, 245, 245)",
        "borderColorEnter": "rgb(204, 204, 204)",
        "borderColorError": "rgb(245, 0, 87)",
        "borderColorFocus": "rgb(0, 188, 212)",
        "borderRadius": "3px",
        "borderWidth": "1px",
        "boxShadow": "0 0 0 transparent",
        "boxShadowDisabled": "0 2px 2px rgba(0, 0, 0, 0.04) inset",
        "boxShadowEnter": "0 2px 2px rgba(0, 0, 0, 0.04) inset",
        "color": "rgb(54, 54, 54)",
        "controlHeight": "2.7em",
        "disabled": "rgb(204, 204, 204)",
        "error": "rgb(245, 0, 87)",
        "fontSize": "1rem",
        "selectArrowColor": "rgb(0, 188, 212)",
        "spinnerColor": "rgb(0, 188, 212)",
        "spinnerDuration": "500ms",
        "transitionDuration": "110ms",
        "transitionProperty": "color, box-shadow, border",
        "transitionTimingFunction": "cubic-bezier(.25, .46, .45, .94)",
      },
    }
  }
>
  <styled.select
    error={false}
    name="mockField"
    onBlur={[Function]}
    onChange={[Function]}
    onFocus={[Function]}
    theme={
      Object {
        "rdf": Object {
          "background": "rgb(255, 255, 255)",
          "backgroundDisabled": "rgb(245, 245, 245)",
          "backgroundEnter": "rgb(255, 255, 255)",
          "backgroundError": "rgb(255, 255, 255)",
          "backgroundFocus": "rgb(255, 255, 255)",
          "borderColor": "rgb(219, 219, 219)",
          "borderColorDisabled": "rgb(245, 245, 245)",
          "borderColorEnter": "rgb(204, 204, 204)",
          "borderColorError": "rgb(245, 0, 87)",
          "borderColorFocus": "rgb(0, 188, 212)",
          "borderRadius": "3px",
          "borderWidth": "1px",
          "boxShadow": "0 0 0 transparent",
          "boxShadowDisabled": "0 2px 2px rgba(0, 0, 0, 0.04) inset",
          "boxShadowEnter": "0 2px 2px rgba(0, 0, 0, 0.04) inset",
          "color": "rgb(54, 54, 54)",
          "controlHeight": "2.7em",
          "disabled": "rgb(204, 204, 204)",
          "error": "rgb(245, 0, 87)",
          "fontSize": "1rem",
          "selectArrowColor": "rgb(0, 188, 212)",
          "spinnerColor": "rgb(0, 188, 212)",
          "spinnerDuration": "500ms",
          "transitionDuration": "110ms",
          "transitionProperty": "color, box-shadow, border",
          "transitionTimingFunction": "cubic-bezier(.25, .46, .45, .94)",
        },
      }
    }
    value={null}
  >
    select field
  </styled.select>
</styled.div>
`;

exports[`<Select /> Should be render disabled control 1`] = `
<styled.div
  showSpinner={false}
  theme={
    Object {
      "rdf": Object {
        "background": "rgb(255, 255, 255)",
        "backgroundDisabled": "rgb(245, 245, 245)",
        "backgroundEnter": "rgb(255, 255, 255)",
        "backgroundError": "rgb(255, 255, 255)",
        "backgroundFocus": "rgb(255, 255, 255)",
        "borderColor": "rgb(219, 219, 219)",
        "borderColorDisabled": "rgb(245, 245, 245)",
        "borderColorEnter": "rgb(204, 204, 204)",
        "borderColorError": "rgb(245, 0, 87)",
        "borderColorFocus": "rgb(0, 188, 212)",
        "borderRadius": "3px",
        "borderWidth": "1px",
        "boxShadow": "0 0 0 transparent",
        "boxShadowDisabled": "0 2px 2px rgba(0, 0, 0, 0.04) inset",
        "boxShadowEnter": "0 2px 2px rgba(0, 0, 0, 0.04) inset",
        "color": "rgb(54, 54, 54)",
        "controlHeight": "2.7em",
        "disabled": "rgb(204, 204, 204)",
        "error": "rgb(245, 0, 87)",
        "fontSize": "1rem",
        "selectArrowColor": "rgb(0, 188, 212)",
        "spinnerColor": "rgb(0, 188, 212)",
        "spinnerDuration": "500ms",
        "transitionDuration": "110ms",
        "transitionProperty": "color, box-shadow, border",
        "transitionTimingFunction": "cubic-bezier(.25, .46, .45, .94)",
      },
    }
  }
>
  <styled.select
    disabled={true}
    error={false}
    name="mockField"
    onBlur={[Function]}
    onChange={[Function]}
    onFocus={[Function]}
    theme={
      Object {
        "rdf": Object {
          "background": "rgb(255, 255, 255)",
          "backgroundDisabled": "rgb(245, 245, 245)",
          "backgroundEnter": "rgb(255, 255, 255)",
          "backgroundError": "rgb(255, 255, 255)",
          "backgroundFocus": "rgb(255, 255, 255)",
          "borderColor": "rgb(219, 219, 219)",
          "borderColorDisabled": "rgb(245, 245, 245)",
          "borderColorEnter": "rgb(204, 204, 204)",
          "borderColorError": "rgb(245, 0, 87)",
          "borderColorFocus": "rgb(0, 188, 212)",
          "borderRadius": "3px",
          "borderWidth": "1px",
          "boxShadow": "0 0 0 transparent",
          "boxShadowDisabled": "0 2px 2px rgba(0, 0, 0, 0.04) inset",
          "boxShadowEnter": "0 2px 2px rgba(0, 0, 0, 0.04) inset",
          "color": "rgb(54, 54, 54)",
          "controlHeight": "2.7em",
          "disabled": "rgb(204, 204, 204)",
          "error": "rgb(245, 0, 87)",
          "fontSize": "1rem",
          "selectArrowColor": "rgb(0, 188, 212)",
          "spinnerColor": "rgb(0, 188, 212)",
          "spinnerDuration": "500ms",
          "transitionDuration": "110ms",
          "transitionProperty": "color, box-shadow, border",
          "transitionTimingFunction": "cubic-bezier(.25, .46, .45, .94)",
        },
      }
    }
    value={null}
  />
</styled.div>
`;

exports[`<Select /> Should be render error text 1`] = `
<styled.div
  showSpinner={false}
  theme={
    Object {
      "rdf": Object {
        "background": "rgb(255, 255, 255)",
        "backgroundDisabled": "rgb(245, 245, 245)",
        "backgroundEnter": "rgb(255, 255, 255)",
        "backgroundError": "rgb(255, 255, 255)",
        "backgroundFocus": "rgb(255, 255, 255)",
        "borderColor": "rgb(219, 219, 219)",
        "borderColorDisabled": "rgb(245, 245, 245)",
        "borderColorEnter": "rgb(204, 204, 204)",
        "borderColorError": "rgb(245, 0, 87)",
        "borderColorFocus": "rgb(0, 188, 212)",
        "borderRadius": "3px",
        "borderWidth": "1px",
        "boxShadow": "0 0 0 transparent",
        "boxShadowDisabled": "0 2px 2px rgba(0, 0, 0, 0.04) inset",
        "boxShadowEnter": "0 2px 2px rgba(0, 0, 0, 0.04) inset",
        "color": "rgb(54, 54, 54)",
        "controlHeight": "2.7em",
        "disabled": "rgb(204, 204, 204)",
        "error": "rgb(245, 0, 87)",
        "fontSize": "1rem",
        "selectArrowColor": "rgb(0, 188, 212)",
        "spinnerColor": "rgb(0, 188, 212)",
        "spinnerDuration": "500ms",
        "transitionDuration": "110ms",
        "transitionProperty": "color, box-shadow, border",
        "transitionTimingFunction": "cubic-bezier(.25, .46, .45, .94)",
      },
    }
  }
>
  <styled.select
    error={true}
    name="mockField"
    onBlur={[Function]}
    onChange={[Function]}
    onFocus={[Function]}
    theme={
      Object {
        "rdf": Object {
          "background": "rgb(255, 255, 255)",
          "backgroundDisabled": "rgb(245, 245, 245)",
          "backgroundEnter": "rgb(255, 255, 255)",
          "backgroundError": "rgb(255, 255, 255)",
          "backgroundFocus": "rgb(255, 255, 255)",
          "borderColor": "rgb(219, 219, 219)",
          "borderColorDisabled": "rgb(245, 245, 245)",
          "borderColorEnter": "rgb(204, 204, 204)",
          "borderColorError": "rgb(245, 0, 87)",
          "borderColorFocus": "rgb(0, 188, 212)",
          "borderRadius": "3px",
          "borderWidth": "1px",
          "boxShadow": "0 0 0 transparent",
          "boxShadowDisabled": "0 2px 2px rgba(0, 0, 0, 0.04) inset",
          "boxShadowEnter": "0 2px 2px rgba(0, 0, 0, 0.04) inset",
          "color": "rgb(54, 54, 54)",
          "controlHeight": "2.7em",
          "disabled": "rgb(204, 204, 204)",
          "error": "rgb(245, 0, 87)",
          "fontSize": "1rem",
          "selectArrowColor": "rgb(0, 188, 212)",
          "spinnerColor": "rgb(0, 188, 212)",
          "spinnerDuration": "500ms",
          "transitionDuration": "110ms",
          "transitionProperty": "color, box-shadow, border",
          "transitionTimingFunction": "cubic-bezier(.25, .46, .45, .94)",
        },
      }
    }
    value={null}
  />
  <styled.div
    theme={
      Object {
        "rdf": Object {
          "background": "rgb(255, 255, 255)",
          "backgroundDisabled": "rgb(245, 245, 245)",
          "backgroundEnter": "rgb(255, 255, 255)",
          "backgroundError": "rgb(255, 255, 255)",
          "backgroundFocus": "rgb(255, 255, 255)",
          "borderColor": "rgb(219, 219, 219)",
          "borderColorDisabled": "rgb(245, 245, 245)",
          "borderColorEnter": "rgb(204, 204, 204)",
          "borderColorError": "rgb(245, 0, 87)",
          "borderColorFocus": "rgb(0, 188, 212)",
          "borderRadius": "3px",
          "borderWidth": "1px",
          "boxShadow": "0 0 0 transparent",
          "boxShadowDisabled": "0 2px 2px rgba(0, 0, 0, 0.04) inset",
          "boxShadowEnter": "0 2px 2px rgba(0, 0, 0, 0.04) inset",
          "color": "rgb(54, 54, 54)",
          "controlHeight": "2.7em",
          "disabled": "rgb(204, 204, 204)",
          "error": "rgb(245, 0, 87)",
          "fontSize": "1rem",
          "selectArrowColor": "rgb(0, 188, 212)",
          "spinnerColor": "rgb(0, 188, 212)",
          "spinnerDuration": "500ms",
          "transitionDuration": "110ms",
          "transitionProperty": "color, box-shadow, border",
          "transitionTimingFunction": "cubic-bezier(.25, .46, .45, .94)",
        },
      }
    }
  >
    error!!
  </styled.div>
</styled.div>
`;

exports[`<Select /> Should be render spinner 1`] = `
<styled.div
  showSpinner={true}
  theme={
    Object {
      "rdf": Object {
        "background": "rgb(255, 255, 255)",
        "backgroundDisabled": "rgb(245, 245, 245)",
        "backgroundEnter": "rgb(255, 255, 255)",
        "backgroundError": "rgb(255, 255, 255)",
        "backgroundFocus": "rgb(255, 255, 255)",
        "borderColor": "rgb(219, 219, 219)",
        "borderColorDisabled": "rgb(245, 245, 245)",
        "borderColorEnter": "rgb(204, 204, 204)",
        "borderColorError": "rgb(245, 0, 87)",
        "borderColorFocus": "rgb(0, 188, 212)",
        "borderRadius": "3px",
        "borderWidth": "1px",
        "boxShadow": "0 0 0 transparent",
        "boxShadowDisabled": "0 2px 2px rgba(0, 0, 0, 0.04) inset",
        "boxShadowEnter": "0 2px 2px rgba(0, 0, 0, 0.04) inset",
        "color": "rgb(54, 54, 54)",
        "controlHeight": "2.7em",
        "disabled": "rgb(204, 204, 204)",
        "error": "rgb(245, 0, 87)",
        "fontSize": "1rem",
        "selectArrowColor": "rgb(0, 188, 212)",
        "spinnerColor": "rgb(0, 188, 212)",
        "spinnerDuration": "500ms",
        "transitionDuration": "110ms",
        "transitionProperty": "color, box-shadow, border",
        "transitionTimingFunction": "cubic-bezier(.25, .46, .45, .94)",
      },
    }
  }
>
  <styled.select
    error={false}
    name="mockField"
    onBlur={[Function]}
    onChange={[Function]}
    onFocus={[Function]}
    theme={
      Object {
        "rdf": Object {
          "background": "rgb(255, 255, 255)",
          "backgroundDisabled": "rgb(245, 245, 245)",
          "backgroundEnter": "rgb(255, 255, 255)",
          "backgroundError": "rgb(255, 255, 255)",
          "backgroundFocus": "rgb(255, 255, 255)",
          "borderColor": "rgb(219, 219, 219)",
          "borderColorDisabled": "rgb(245, 245, 245)",
          "borderColorEnter": "rgb(204, 204, 204)",
          "borderColorError": "rgb(245, 0, 87)",
          "borderColorFocus": "rgb(0, 188, 212)",
          "borderRadius": "3px",
          "borderWidth": "1px",
          "boxShadow": "0 0 0 transparent",
          "boxShadowDisabled": "0 2px 2px rgba(0, 0, 0, 0.04) inset",
          "boxShadowEnter": "0 2px 2px rgba(0, 0, 0, 0.04) inset",
          "color": "rgb(54, 54, 54)",
          "controlHeight": "2.7em",
          "disabled": "rgb(204, 204, 204)",
          "error": "rgb(245, 0, 87)",
          "fontSize": "1rem",
          "selectArrowColor": "rgb(0, 188, 212)",
          "spinnerColor": "rgb(0, 188, 212)",
          "spinnerDuration": "500ms",
          "transitionDuration": "110ms",
          "transitionProperty": "color, box-shadow, border",
          "transitionTimingFunction": "cubic-bezier(.25, .46, .45, .94)",
        },
      }
    }
    value={null}
  />
  <styled.div
    theme={
      Object {
        "rdf": Object {
          "background": "rgb(255, 255, 255)",
          "backgroundDisabled": "rgb(245, 245, 245)",
          "backgroundEnter": "rgb(255, 255, 255)",
          "backgroundError": "rgb(255, 255, 255)",
          "backgroundFocus": "rgb(255, 255, 255)",
          "borderColor": "rgb(219, 219, 219)",
          "borderColorDisabled": "rgb(245, 245, 245)",
          "borderColorEnter": "rgb(204, 204, 204)",
          "borderColorError": "rgb(245, 0, 87)",
          "borderColorFocus": "rgb(0, 188, 212)",
          "borderRadius": "3px",
          "borderWidth": "1px",
          "boxShadow": "0 0 0 transparent",
          "boxShadowDisabled": "0 2px 2px rgba(0, 0, 0, 0.04) inset",
          "boxShadowEnter": "0 2px 2px rgba(0, 0, 0, 0.04) inset",
          "color": "rgb(54, 54, 54)",
          "controlHeight": "2.7em",
          "disabled": "rgb(204, 204, 204)",
          "error": "rgb(245, 0, 87)",
          "fontSize": "1rem",
          "selectArrowColor": "rgb(0, 188, 212)",
          "spinnerColor": "rgb(0, 188, 212)",
          "spinnerDuration": "500ms",
          "transitionDuration": "110ms",
          "transitionProperty": "color, box-shadow, border",
          "transitionTimingFunction": "cubic-bezier(.25, .46, .45, .94)",
        },
      }
    }
  />
</styled.div>
`;
